rand_core/src/impls.rs

```diff
@@ -160,6 +160,49 @@ pub fn next_u64_via_fill<R: RngCore + ?Sized>(rng: &mut R) -> u64 {
     u64::from_le_bytes(buf)
 }
 
+/// Implement [`TryRngCore`] for a type implementing [`RngCore`].
+///
+/// Ideally, `rand_core` would define blanket impls for this, but they conflict with blanket impls
+/// for `&mut R` and `Box<R>`, so until specialziation is stabilized, implementer crates
+/// have to implement `TryRngCore` directly.
+#[macro_export]
+macro_rules! impl_try_rng_from_rng_core {
+    ($t:ty) => {
+        impl $crate::TryRngCore for $t {
+            type Error = core::convert::Infallible;
+
+            #[inline]
+            fn try_next_u32(&mut self) -> Result<u32, Self::Error> {
+                Ok(self.next_u32())
+            }
+
+            #[inline]
+            fn try_next_u64(&mut self) -> Result<u64, Self::Error> {
+                Ok(self.next_u64())
+            }
+
+            #[inline]
+            fn try_fill_bytes(&mut self, dst: &mut [u8]) -> Result<(), Self::Error> {
+                self.fill_bytes(dst);
+                Ok(())
+            }
+        }
+    };
+}
+
+/// Implement [`TryCryptoRng`] for a type implementing [`CryptoRng`].
+///
+/// Ideally, `rand_core` would define blanket impls for this, but they conflict with blanket impls
+/// for `&mut R` and `Box<R>`, so until specialziation is stabilized, implementer crates
+/// have to implement `TryRngCore` and `TryCryptoRng` directly.
+#[macro_export]
+macro_rules! impl_try_crypto_rng_from_crypto_rng {
+    ($t:ty) => {
+        $crate::impl_try_rng_from_rng_core!($t);
+        impl $crate::TryCryptoRng for $t {}
+    };
+}
```

Comment:
Caveat: the macro name makes it sound like it requires that `CryptoRng` is implemented. There is no check for this. 
Caveat: the call to `impl_try_rng_from_rng_core` is not documented.

We could just have crates write the impl directly, but I'd prefer not to (to make it easier to add a blanket impl when specialization is available).

Possible solution: add `#[doc(hidden)] fn check_is_crypto_rng<R: CryptoRng>() {}` to `rand_core` and a static assertion using this over `$t` to the macro.

Reply:
I've used an anonymous `const` block instead:
```rust
```